src/functions.php

```diff
@@ -377,13 +386,128 @@ function upload(string $xml, $config)
         'PhonebookImportFile' => [
             'type' => 'text/xml',
             'filename' => 'updatepb.xml',
-            'content' => $xml,
+            'content' => $xmlPhonebook->asXML(), // convert XML object to XML string
         ]
     ];
 
-    $result = $fritz->postFile($formfields, $filefields); // send the command
-
+    $result = $fritz->postFile($formfields, $filefields); // send the command to store new phonebook
     if (strpos($result, 'Das Telefonbuch der FRITZ!Box wurde wiederhergestellt') === false) {
         throw new \Exception('Upload failed');
     }
 }
+
+
+/**
+ * Downloads the phone book from Fritzbox
+ *
+ * @param   Api   $fritz
+ * @param   array $config
+ * @return  SimpleXMLElement with the old existing phonebook
+ */
+function downloadPhonebook(Api $fritz, array $config)
+{
+    $formfields = [
+        'PhonebookId' => $config['phonebook']['id'],
+        'PhonebookExportName' => $config['phonebook']['name'],
+        'PhonebookExport' => "",
+    ];
+    $result = $fritz->postFile($formfields, []); // send the command to load existing phone book
+    if (substr($result, 0, 5) !== "<?xml") {
+        error_log("ERROR: Could not load phonebook with ID=".$config['phonebook']['id']);
+        return new SimpleXMLElement("<xml><empty /></xml>");
+    }
+    $xmlPhonebook = simplexml_load_string($result);
+    return $xmlPhonebook;
+}
+
+
+/**
+ * Get quickdial and vanity special attributes from given XML phone book
+ *
+ * @param   SimpleXMLElement    $xmlPhonebook
+ * @return  mixed               [] or map with {phonenumber@CardDavUID}=>SimpleXMLElement-Attributes
```

Comment:
type is array then

Reply:
2 x OK.

Reply:
phpstan did'nt like the "array" type here...

Reply:
Maybe it would work if you declare it `array<string, object>`. Seems stan recognizes the map.

Reply:
I'll give it a shot.  :-)

Reply:
Seems to work! https://travis-ci.org/andig/carddav2fb/builds/510572172 is green...